src/Statistics.jl

```diff
@@ -0,0 +1,135 @@
+using Random: shuffle!
+using StatsBase: AbstractWeights, weights, values
+@doc doc"""
+    mean(M,x, weights=1/n*ones(n);, stop_iter=100, kwargs... )
+
+computes the Riemannian center of mass also known as Karcher mean of the vector
+`x` of `n` points on the [`Manifold`](@ref) `M`. This function
+uses the gradient descent scheme. Optionally one can provide
+weights $w_i$ for the weighted Riemannian center of mass.
+The general formula to compute the minimizer reads
+````math
+\argmin_{y\in\mathcal M} \frac{1}{2}\sum_{i=1}^n w_i\mathrm{d}_{\mathcal M}^2(y,x_i),
+````
+where $\mathrm{d}_{\mathcal M}$ denotes the Riemannian [`distance`](@ref).
+
+Optionally you can provide `x0`, the starting point (by default set to the first
+data point). `stop_iter` denotes the maximal number of iterations to perform and
+the `kwargs...` are passed to [`isapprox`](@ref) to stop, when the inimal change
+between two iterates is small. For more stopping criteria
+check the [`Manopt.jl`](https://manoptjl.org) package and use a solver therefrom.
+
+The algorithm is further described in 
+> Afsari, B; Tron, R.; Vidal, R.: On the Convergence of Gradient
+> Descent for Finding the Riemannian Center of Mass,
+> SIAM Journal on Control and Optimization (2013), 51(3), pp. 2230–2260, 
+> doi: [10.1137/12086282X](https://doi.org/10.1137/12086282X),
+> arxiv: [1201.0925](https://arxiv.org/abs/1201.0925">1201.0925)
+"""
+function mean(M::Manifold, x::AbstractVector, w::AbstractWeights = weights(ones(length(x)) / length(x)); x0 = x[1], kwargs...)
+    y = similar_result(M, mean, x0)
+    copyto!(y,x0)
+    return mean!(M, y, x, w; kwargs...)
+end
+mean(M::Euclidean, x::AbstractVector, w::AbstractWeights) = mean(x,w)
+mean(M::Euclidean, x::AbstractVector) = mean(x)
+@doc doc"""
+    mean!(M,y,x,w)
+
+computes the [`mean`](@ref) in-place in `y` where the initial value of `y` is
+the starting point of the algorithm.
+"""
+function mean!(M::Manifold, y, x::AbstractVector, w::AbstractWeights = weights(ones(length(x)) / length(x));
+            stop_iter=100,
+            kwargs...
+        ) where {T}
+    yold = similar_result(M, mean, y)
+    copyto!(yold,y)
+    v = zero_tangent_vector.(Ref(M), fill(y,length(x)))
+    for i=1:stop_iter
+        copyto!(yold,y)
+        log!.(Ref(M), v, Ref(yold), x)
+        exp!(M, y, yold, sum( values(w).*v ) / 2 )
```

Comment:
Could declare a `vmean` above and then do
```suggestion
        mean!(vmean, v, w)
        exp!(M, y, yold, vmean)
```

Reply:
The 2 stems from regularising the step size, the `n` is hidden in the weights (which sum op to one, but I till correct that as you proposed below).

Reply:
Concerning the vmean – you mean in the sense of a mean on the tangent vectors? hm, yes one could do that.

Reply:
Yes I like calling `vmean` on the tangent vectors better. It avoids any funny business with the weights, since we let StatsBase handle the usual weighted mean in the tangent space.
I still don't understand the 2, but I trust you. 🙂

Reply:
Everything works as intended now.

Reply:
I don't think so? There's no `w.sum` anywhere.

Reply:
Then you might have finally have lost me here and have to explain much more. I worked on this quite a while so I think it does what it should. If you think it does not – please provide better code.

Reply:
Do you mind if I push some changes with tests later today? You can always revert if you disagree.

Reply:
Feel free to do so, when I ask for code, I am completely fine with a push.